### ISSUE TITLE
Mock creation throws NPE when class extends spark Logging and calls log

### ISSUE THREAD

**Issue by gongfunaicha** [paulbutcher/ScalaMock#247]
*If you want to discuss a new feature, please ignore/clear this form.*

### ScalaMock Version (e.g. 3.5.0)
3.5.0

### Scala Version (e.g. 2.12)
2.11

### Runtime (JVM or JS)
JVM

### Please describe the expected behavior of the issue
Mock should be created successfully. No exceptions thrown.

### Please provide a description of what actually happens
NullPointerException is thrown during runtime.

```
An exception or error caused a run to abort. 
java.lang.NullPointerException
	at FooTest$$anon$1.log(FooTest.scala:6)
	at Foo.<init>(FooTest.scala:10)
	at FooTest$$anon$1.<init>(FooTest.scala:6)
	at FooTest.<init>(FooTest.scala:6)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at org.scalatest.tools.Runner$.genSuiteConfig(Runner.scala:1422)
	at org.scalatest.tools.Runner$$anonfun$31.apply(Runner.scala:1236)
	at org.scalatest.tools.Runner$$anonfun$31.apply(Runner.scala:1235)
	at scala.collection.immutable.List.map(List.scala:284)
	at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:1235)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1011)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1010)
	at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:1500)
	at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1010)
	at org.scalatest.tools.Runner$.run(Runner.scala:850)
	at org.scalatest.tools.Runner.run(Runner.scala)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.runScalaTest2(ScalaTestRunner.java:131)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.main(ScalaTestRunner.java:28)
```

### Reproducible Test Case

Please provide a minimised code snippet that fails, ideally, written as a failing test case in ScalaTest.
This will help us a lot in diagnosing the problem and work on a fix.
If the issue is more complex or requires configuration, please provide a link to a project on Github that reproduces the issue.

```scala
import org.apache.spark.internal.Logging
import org.scalamock.scalatest.MockFactory
import org.scalatest.FlatSpec

class FooTest extends FlatSpec with MockFactory {
  mock[Foo]
}

class Foo extends Logging {
  log.info("foo log")
}
```

**Comment by barkhorn** [comment]
The stacktrace gives you a clue what happens here: `Foo.<init>(FooTest.scala:10)`
Your constructor of `Foo` is relying on the actual implementation of `Logging` but the whole purpose of a mock is to not use the original implementation at all. So this cannot work.

As you can see in the spark codebase, log is a protected method, so a mock will not include this code. https://github.com/apache/spark/blob/0e31a6f25e0263b144255a6630e1d381fe2d27a7/core/src/main/scala/org/apache/spark/internal/Logging.scala#L44

In other words, this is not a bug, but an invalid use-case for mocks.
I think there are frameworks which support partial mocking, but for ScalaMock, this is definitely out of scope.

**Comment by gongfunaicha** [comment]
@barkhorn 
Hey Philipp,

Thanks for the response. Could you explain a bit more on why this is an invalid use-case for mocks? The situation is that I don't need the logging to be executed but it is in the constructor of the class which I'd like to mock. Does this mean that I cannot use ScalaMock if I'd like to mock a class which calls a protected method of super class in the constructor?

**Comment by barkhorn** [comment]
Let me rephrase this a little differently then:
1. mocks are an (anonymous) subclass of the type to mock
2. mocks tear off the original implementation and delegate to call handlers for the purpose of mocking calls (with a few exceptions, e.g. private methods)
3. per JVM convention, subclass constructors will call their baseclass constructor first

`Foo`s constructor depends on the concrete implementation of `log`, but in a mock this will delegate to a call handler, and that could not be initialised yet due to constructor precedence. So it generates an NPE.
Calling overridable methods in a constructor is bad design, as that can lead to partially constructed objects like in this case.
A possible way around this would be to use objenesis to avoid calling constructors, I am experimenting with that. Right now though, because of the way the JVM works, you will need to find a workaround.
